### ISSUE TITLE
Inline ASK queries are not updated automatically

### ISSUE THREAD

**Issue by HueyNemud** [SemanticMediaWiki/SemanticMediaWiki#4974]
Hello,

Since I upgraded my wiki to MW 1.35.1 & SMW 3.2.3 (with Semantic Result Formats 3.2.0) I am facing an issue with automatic update of inline #ask queries.

I have a set of pages belonging to two categories, `[[Category:Monographie]]` and `[[Category:À propos de ce site]]`, and an index page containing this query: 
```
{{#ask:[[Catégorie:Monographie]][[Catégorie:À propos du site]]|format=ul}}
```

If I remove/add one of those categories from a page so that it should affect the query, its result is not automatically updated. Even with a manual purge.

The only way to get the results to be updated is to either restart Apache, or edit & save the index page.

The same seems to happen with all #ask queries that have been created before upgrading to MW 1.35.1/SMW 3.2.3, but if I add a completely new category to pages (new ones or existing) and create a similar query then it is automatically updated as intended.

What's wierd is that if I run the SQL query I get from `format=debug`, I get the correct list of pages.

Note that I have query cache & query update enabled in LocalSettings.php : 
```php
$smwgEnabledQueryDependencyLinksStore = true;
$smwgMainCacheType = CACHE_ACCEL;
$smwgQueryResultCacheType = CACHE_ANYTHING;
$smwgQueryResultCacheLifetime = 60 * 60 * 24 * 30; # Keep query results for 1 month.
$smwgQueryResultCacheRefreshOnPurge = true; # Refresh query cache on manual purge
```
Yet even with those lines commented out nothing changes.

Disabling Semantic Result Format has no effect either.

Maybe this is related to #4937 ?

Other infos:
- Debian 9
- PHP 7.4.3
- Mysql 8
- Object caching with APCu

**Comment by kghbln** [comment]
It works for me without issues for the following setup:

**Software**
- MediaWiki | 1.31.14 (f85ff27)22:33, 8. Apr. 2021
- PHP | 7.3.27-1~deb10u1 (apache2handler)
- MariaDB | 10.3.27-MariaDB-0+deb10u1
- Semantic MediaWiki | 3.2.3 (5619c2d) 00:40, 30. Mär. 2021

**Configuration**
```php
$wgMainCacheType = 'redis';

$smwgEnabledQueryDependencyLinksStore = true;
$smwgQueryResultCacheType = CACHE_MEMCACHED; // actually I should switch to redis here, too

$smwgMainCacheType = CACHE_ANYTHING; // default, thus not explicitly set
$smwgQueryResultCacheLifetime = 60 * 60 * 24 * 7; // default, thus not explicitly set
$smwgQueryResultCacheRefreshOnPurge = true; // default, thus not explicitly set
```

The job queue is cleared every second minute.

**Comment by HueyNemud** [comment]
Hi,

Unfortunately for hosting reasons I cannot use external tools like Redis, I'm stuck with APCu.

I investigated the issue a bit further and disabled all caches and all non default extensions.
```php 
# $wgMainCacheType and $smwgQueryResultCacheType default values are CACHE_NONE
$smwgMainCacheType = CACHE_NONE;
$smwgEnabledQueryDependencyLinksStore = true;
```

Page A is in categories `[[Catégorie:Monographie]][[Catégorie:À propos du site]]`, page B contains the embedded #ask query from my initial post.

If I remove A from both categories, one htmlCacheUpdate job is fired and processed, but B's query is not updated until I purge the page.
If I now add back A in categories `Monographie` and `À propos du site`, one recentChangesUpdate job is processed, but B is not updated.

Still, the debug SQL query always give the correct result.
Maybe this is more about MW html cache invalidation than the embedded query itself ?

**Comment by kghbln** [comment]
Well, the biggest difference I see here is the MediaWiki version. Perhaps you could try on sandbox and see how that wiki reacts.

**Comment by thomas-topway-it** [comment]
For us it is related to the fact that SMW, in combination to our extension, does not save the page category among the data, so  the category does not appear using

```wiki/Special:Browse```

(so that if included in an ask query, the page does not appear among the results)
So it is an issue on our end, but I have still to verify what is the precise function in our extension causing this behavior

**Comment by thomas-topway-it** [comment]
so, on our side, I have found out that the conflicting hook is 
```
ParserAfterTidy
```
it seems that as long as you use this hook, the category is not included among the page data from SMW, @kghbln, can you check on your side ?
(Thomas)

**Comment by gesinn-it** [comment]
@thomas-topway-it did I get you right that you have an own/custom extension that is using the `ParserAfterTidy` hook and you have the issues described as long as your extension is loaded? Does everything work as expected, when you temp. disable your extension?

**Comment by thomas-topway-it** [comment]
@gesinn-it  yes, as long as you enable the ```ParserAfterTidy``` hook the category is not anymore being recorded by SMW. When you disable it (and you save the page again) it works as expected

**Comment by gesinn-it** [comment]
@thomas-topway-it not sure if I understood what you mean by "enable the ... hook". A hook allows custom code to be executed when some defined event (such as saving a page or a user logging in) occurs. `ParserAfterTidy` is a MediaWiki hook, so you can not "disable" or "enable" it. You can "disable" (not load) extensions using this hook.

If other extensions using `ParserAfterTidy` [1] do not cause the issue, it might be an issue with your extension.

SMW itself is also using `ParserAfterTidy`, could also be an issue there: https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/91546f4c02af0409f97bacecd142a6170d78cde2/src/MediaWiki/Hooks/ParserAfterTidy.php. There is also a test https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/91546f4c02af0409f97bacecd142a6170d78cde2/tests/phpunit/Unit/MediaWiki/Hooks/ParserAfterTidyTest.php. Can you try to run this test and see if it fails in your environment?

[1] https://www.mediawiki.org/wiki/Category:ParserAfterTidy_extensions

**Comment by thomas-topway-it** [comment]
with enabling/disabling I mean adding/removing the hook to/from the extension.
That also occurs when the function attached to the hook in our extension is empty, so I thought it is an issue of SMW.
I will do the test whenever possible, thanks
